src/video_core/amdgpu/resource.h

```diff
@@ -20,6 +21,77 @@ enum class CompSwizzle : u32 {
     Alpha = 7,
 };
 
+struct CompMapping {
+    CompSwizzle r : 3;
+    CompSwizzle g : 3;
+    CompSwizzle b : 3;
+    CompSwizzle a : 3;
+
+    auto operator<=>(const CompMapping& other) const = default;
+
+    template <typename T>
+    [[nodiscard]] std::array<T, 4> Apply(const std::array<T, 4>& data) const {
+        return Apply<T>([&data](const u32 index) { return data[index]; },
+                        [](const u32 imm) { return T(imm); });
+    }
+
+    template <typename T>
+    [[nodiscard]] std::array<T, 4> Apply(const std::function<T(u32)>& get_value,
+                                         const std::function<T(u32)>& get_immediate) const {
```

Comment:
```suggestion
    [[nodiscard]] std::array<T, 4> Apply(const auto& get_value, const auto& get_immediate) const {
```
I'm not sure how well std::function is optimized by compiler but it might do heap allocation + type erasure. Using auto ensures we pass a lambda which will be inlined by compiler